### PR TITLE
V2E-4145: integration ids

### PR DESCRIPTION
**Pull request by dallonf**
* **Please check if the PR fulfills these requirements**
    + [ ] Tests for the changes have been added (for bug fixes / features)
       + N/A (not enough existing test infrastructure to build on)
    + [x] Docs have been added / updated (for bug fixes / features)
    + [x] Addresses JIRA issue V2E-4145

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

New feature!

* **What is the current behavior?**

"Integration" isn't supported

* **What is the new behavior (if this is a feature change)?**

You can pass `--integration` when publishing or updating a block.

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

Don't believe so!

### CODE REVIEW

**Review by divinentd — APPROVED** [review]

File: src/utils/validation.ts [hunk 1]
```diff
@@ -76,3 +102,18 @@ export const validateBlockDirectory = (): void => {
         process.exit(1);
     }
 };
+
+export const validateIntegration = (
```
Comment (greg-murray-volusion) [inline src/utils/validation.ts:106]: I think we need to change the name of this one or have a separate function for translating the name to id. What do you think?

File: src/utils/validation.ts [hunk 2]
```diff
@@ -76,3 +102,18 @@ export const validateBlockDirectory = (): void => {
...
+    integrationName: string | undefined
+): number | undefined => {
+    if (!integrationName) {
+        return undefined;
+    }
+
+    const integration = INTEGRATIONS[integrationName!];
+    if (!integration) {
+        logError("Supported integrations are 'volt', 'standard', and 'v1'.");
```
Comment (greg-murray-volusion) [inline src/utils/validation.ts:115]: Unsupported category error puts each item on a separate line. Maybe map them from `INTEGRATIONS` keys sorted by ids but only listing the name.

File: src/utils/validation.ts [hunk 3]
```diff
@@ -76,3 +102,18 @@ export const validateBlockDirectory = (): void => {
...
 };
+
+export const validateIntegration = (
+    integrationName: string | undefined
+): number | undefined => {
+    if (!integrationName) {
+        return undefined;
+    }
+
+    const integration = INTEGRATIONS[integrationName!];
```
[the remark this answers is not in the document]
Comment (dallonf) [reply src/utils/validation.ts:113]: Good idea!

File: src/utils/validation.ts (hunk 1 above)
Comment (dallonf) [reply src/utils/validation.ts:106]: I think splitting out the functions would be tricky, so I'll try a name of `integrationIdFromName`

File: src/utils/validation.ts (hunk 2 above)
Comment (dallonf) [reply src/utils/validation.ts:115]: Good idea! 
image

File: src/index.ts [hunk 4]
```diff
@@ -141,12 +142,16 @@ program
         "-c, --category [category]",
         "The Category name that best fits this block"
     )
+    .option(
+        "-i, --integration [integration]",
+        "The integration this block is built for. Options are 'volt', 'standard', and 'v1'. Defaults to 'standard'."
```
[the remark this answers is not in the document]
Comment (dallonf) [reply src/index.ts:147]: Something doesn't feel quite right to me about hardcoding a number of tabs to try to match commander's indentation, but I did go ahead and connect the docs strings to the configuration data so we only have to update it in one place going forward.

**Review by greg-murray-volusion — APPROVED** [review]

File: src/commands/publish.ts [hunk 5]
```diff
@@ -31,36 +31,50 @@ import {
...
+    name,
+    category,
+    categories,
+    integrationName,
+}: {
+    name: string | null;
+    category: string;
+    categories?: string[];
+    integrationName?: string;
+}): Promise<void> => {
```
Comment (cameron-jones-volusion) [inline src/commands/publish.ts:46]: **Suggestion:** naming this interface could make it a little more readable. Great to have types here either way, though!

File: src/constants.ts [hunk 6]
```diff
@@ -7,3 +7,16 @@ export const RC_FILE_PATH = `${homedir()}/.volusionrc`;
 export const THUMBNAIL_PATH = "thumbnail.png";
 export const USER_DEFINED_BLOCK_CONFIG_FILE = "./dist/defaultConfig.json";
 export const ELEMENT_VERSION = packageFile.version;
+
+export const INTEGRATIONS: Record<string, { id: number } | undefined> = {
```
Comment (cameron-jones-volusion) [inline src/constants.ts:11]: **Suggestion:** name this type for more clarity.

File: src/constants.ts [hunk 7]
```diff
@@ -7,3 +7,16 @@ export const RC_FILE_PATH = `${homedir()}/.volusionrc`;
...
+    volt: {
+        id: 1,
+    },
+    // tslint:disable-next-line: object-literal-sort-keys
```
Comment (cameron-jones-volusion) [inline src/constants.ts:15]: **Suggestion:** You can get around this eslint rule by adding line breaks between keys that don't need to be sorted together.

File: src/index.ts [hunk 8]
```diff
@@ -141,12 +146,20 @@ program
         "-c, --category [category]",
         "The Category name that best fits this block"
     )
+    .option(
+        "-i, --integration [integration]",
+        `The integration this block is built for. Options are: ${Object.keys(
+            INTEGRATIONS
+        )
+            .map((x) => `"${x}"`)
+            .join(", ")}. Defaults to 'standard'.`
```
Comment (cameron-jones-volusion) [inline src/index.ts:155]: **Suggestion:** extract the integration-listing code to a named function:
```typescript
`The integration this block is built for. Options are: ${listIntegrations()}. Defaults to 'standard.'`
```

File: src/index.ts [hunk 9]
```diff
@@ -180,7 +193,12 @@ program
                 exit(1);
             }
             if (category) {
-                publish(name, category, categories);
+                publish({
+                    name,
+                    category,
+                    categories,
+                    integrationName: integration,
+                });
```
Comment (cameron-jones-volusion) [inline src/index.ts:201]: love the change to an options object here.

File: src/index.ts [hunk 10]
```diff
@@ -226,11 +248,22 @@ program
...
         "The Category name that best fits this block"
     )
-    .action(({ togglePublic, unminified, category }) => {
+    .option(
+        "-i, --integration [integration]",
+        `The integration this block is built for. Options are: ${Object.keys(
+            INTEGRATIONS
+        )
+            .map((x) => `"${x}"`)
+            .join(", ")}.`
```
Comment (cameron-jones-volusion) [inline src/index.ts:257]: another case for a function to list the integrations

File: src/utils/validation.ts [hunk 11]
```diff
@@ -76,3 +89,24 @@ export const validateBlockDirectory = (): void => {
...
+    }
+
+    const integration = INTEGRATIONS[integrationName!.toLowerCase()];
+    if (!integration) {
+        logError(
+            `\n"${integrationName}" is not a valid integration. The options are:\n${Object.keys(
+                INTEGRATIONS
+            )
+                .map((x) => `\t- ${x}`)
+                .join("\n")}`
```
Comment (cameron-jones-volusion) [inline src/utils/validation.ts:107]: one more place for that integration-listing function.

**Review by cameron-jones-volusion — APPROVED** [review]
Looks good!

Left a few refactoring suggestions, you can take them or leave them. :)

File: src/commands/publish.ts (hunk 5 above)
Comment (dallonf) [reply src/commands/publish.ts:46]: I think in this  case, having the types and function parameters close together is probably the most readable option.

File: src/index.ts (hunk 8 above)
Comment (dallonf) [reply src/index.ts:155]: Couldn't find a good place for a function to live, but I did extract the value and name it, which helps with clutter!

File: src/utils/validation.ts (hunk 11 above)
Comment (dallonf) [reply src/utils/validation.ts:107]: This is actually different enough from the ones in `index` that I think it makes sense to be separate, but I pulled it out into a variable for more clarity. Sometimes I can get a bit carried away with my one-liners :P

**Pull request merged**